sophiread/SophireadCLI/src/sophiread_core.cpp

```diff
@@ -0,0 +1,382 @@
+ */
+void timedLocateTimeStamp(std::vector<TPX3> &batches, const std::vector<char> &rawdata) {
+  auto start = std::chrono::high_resolution_clock::now();
+  unsigned long tdc_timestamp = 0;
+  unsigned long long gdc_timestamp = 0;
+  unsigned long timer_lsb32 = 0;
+  for (auto &tpx3 : batches) {
+    updateTimestamp(tpx3, rawdata, tdc_timestamp, gdc_timestamp, timer_lsb32);
+  }
+  auto end = std::chrono::high_resolution_clock::now();
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+  spdlog::info("Locate all timestamps: {} s", elapsed / 1e6);
+}
+
+/**
+ * @brief Timed hits extraction and clustering via multi-threading.
+ *
+ * @param[in, out] batches
+ * @param[in] rawdata
+ * @param[in] config
+ */
+void timedProcessing(std::vector<TPX3> &batches, const std::vector<char> &raw_data, const IConfig &config) {
+  auto start = std::chrono::high_resolution_clock::now();
+  tbb::parallel_for(tbb::blocked_range<size_t>(0, batches.size()), [&](const tbb::blocked_range<size_t> &r) {
+    // Define ABS algorithm with user-defined parameters for each thread
+    auto abs_alg_mt =
+        std::make_unique<ABS>(config.getABSRadius(), config.getABSMinClusterSize(), config.getABSSpiderTimeRange());
+
+    for (size_t i = r.begin(); i != r.end(); ++i) {
+      auto &tpx3 = batches[i];
+      extractHits(tpx3, raw_data);
+
+      abs_alg_mt->reset();
+      abs_alg_mt->set_method("centroid");
+      abs_alg_mt->fit(tpx3.hits);
+
+      tpx3.neutrons = abs_alg_mt->get_events(tpx3.hits);
+    }
+  });
+  auto end = std::chrono::high_resolution_clock::now();
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+  spdlog::info("Process all hits -> neutrons: {} s", elapsed / 1e6);
+}
+
+/**
+ * @brief Timed save hits to HDF5.
+ *
+ * @param[in] out_hits
+ * @param[in] hits
+ */
+void timedSaveHitsToHDF5(const std::string &out_hits, std::vector<TPX3> &batches) {
+  auto start = std::chrono::high_resolution_clock::now();
+  // move all hits into a single vector
+  std::vector<Hit> hits;
+  for (const auto &tpx3 : batches) {
+    auto tpx3_hits = tpx3.hits;
+    hits.insert(hits.end(), tpx3_hits.begin(), tpx3_hits.end());
+  }
+  // save hits to HDF5 file
+  saveHitsToHDF5(out_hits, hits);
+  auto end = std::chrono::high_resolution_clock::now();
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+  spdlog::info("Save hits to HDF5: {} s", elapsed / 1e6);
+}
+
+/**
+ * @brief Timed save events to HDF5.
+ * 
+ * @param[in] out_events
+ * @param[in] batches
+ */
+void timedSaveEventsToHDF5(const std::string &out_events, std::vector<TPX3> &batches) {
+  auto start = std::chrono::high_resolution_clock::now();
+  // move all events into a single vector
+  std::vector<Neutron> events;
+  for (const auto &tpx3 : batches) {
+    auto tpx3_events = tpx3.neutrons;
+    events.insert(events.end(), tpx3_events.begin(), tpx3_events.end());
```

Comment:
Same comment as before about `std::copy`

Reply:
switched to use `const auto` and `&` to avoid using local copy